sdk/metric/internal/sum.go

```diff
@@ -0,0 +1,71 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package internal // import "go.opentelemetry.io/otel/sdk/metric/internal"
+
+import (
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+)
+
+// sumAgg summarizes a set of measurements as their arithmetic sum.
+type sumAgg[N int64 | float64] struct {
+	// zero value used for the base of all new sums.
+	newFunc NewAtomicFunc[N]
+
+	// map[attribute.Set]Atomic[N]
+	current sync.Map
+}
```

Comment:
I think we are going to have better results if we just use a single lock over two atomic operations (both a load from the map, and a load from the register).  I will put together a branch that would show this off.

```suggestion
type sumAgg[N int64 | float64] struct {
	mu sync.Mutex
	values map[*attribute.Set]N
}
```

Reply:
Here is what the change looks like: https://github.com/MadVikingGod/opentelemetry-go/blob/mvg/new_sdk/impl-sum/sdk/metric/internal/sum.go

Also check out the benchmarks I added. 

Results: https://gist.github.com/MadVikingGod/1d48ff07ee87e3c66acdc33b895d6b3d